### PR TITLE
Jetpack Backup: refactor Jetpack credentials tests validation

### PR DESCRIPTION
**Pull request by Initsogar**
Resolves https://github.com/Automattic/jetpack-backup-team/issues/280

The idea is to refactor our current Jetpack credentials validation in Calypso, so we could introduce a new variable to that validation: restore preflight tests, which we have been introducing in https://github.com/Automattic/wp-calypso/pull/85260 and https://github.com/Automattic/wp-calypso/pull/85786.

Basically, we want to consider some rules before disabling the restore buttons and prompting the user to add credentials. For example, if the preflight tests fail and the site has no credentials configured, we require disabling the restore buttons. But if the site has credentials and are still valid, we should enable the restore. For more details, check the comments and changes made on `getDoesRewindNeedCredentials` selector. 

For now, all preflights tests are failing, so there should not be any impact on final users, as the logic will still consider if the site has credentials and whether they are still valid or not.
  
## Proposed Changes
* Add `hasJetpackCredentials` selectors and its tests
* Prevent credentials test when a site has no credentials.
* Refactor `getDoesRewindNeedCredentials()` selector by introducing some validations like checking if site is atomic and if credentials are invalid and the preflight status.
* Refactor `canRestoreSite()` selector by removing atomic site and credentials test check, given it is available in `getDoesRewindNeedCredentials()`.
* Refactor `BackupStatus` component to remove redundant validations.

We implemented a feature flag for the preflight tests that is only available on development environment for the moment. We need to ensure this changes aren't affecting any real user, so **I'd suggest to test it using the Calypso and Jetpack Cloud Live Branch** (that environment has the feature flag disabled). This means, the preflight test will always be `FAILED`, and the decision of displaying the credentials banner or enabling restore buttons will depend on: site has credentials + credentials are still valid.

| Credentials banner | Restore buttons |
|---|---|
| CleanShot 2024-01-01 at 12 54 53@2x | CleanShot 2024-01-01 at 12 56 18@2x |

### Self-hosted without credentials
- Validate the site has no credentials configured
- Ensure Jetpack Settings page says `Disconnected` and prompt you to enter credentials.
- Ensure all restore buttons are disabled in the following pages (Calypso Blue and Jetpack Cloud):
    - /backup/
    - /activity-log/
- Ensure you see the banner asking credentials in /backup/

**Bonus test:** ensure we are not calling `/rewind/credentials/test?role=main` endpoint when navigating to the backup page in a site without credentials. This is a bugfix.

### Self-hosted with credentials - valid credentials
- Validate the site has credentials configured
- Ensure the credentials are valid, you can check in RWDB.
- Ensure Jetpack Settings page says `Connected` and doesn't prompt you to enter credentials.
- Ensure all restore buttons are enabled, and no banner asking credentials are presented in the following pages (Calypso Blue and Jetpack Cloud)
    - /backup/
    - /activity-log/

### Self-hosted with credentials - invalid credentials
- Validate the site has credentials configured
- Ensure the credentials are no longer valid, you can check in RWDB.
    - If you own the server, you can change the password or disable the user temporarily.
- Ensure all restore buttons are disabled in the following pages (Calypso Blue and Jetpack Cloud):
    - /backup/
    - /activity-log/
- Ensure you see the banner asking credentials in /backup/

### Atomic site & Pressable
- Ensure Jetpack Settings page says `Connected` and doesn't prompt you to enter credentials.
- Ensure all restore buttons are disabled in the following pages (Calypso Blue and Jetpack Cloud):
    - /backup/
    - /activity-log/
 
### Unit tests
Ensure unit tests are passing:
`yarn run test-client client/state/jetpack/credentials client/state/rewind`

## Pre-merge Checklist

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

### CODE REVIEW

File: client/state/selectors/get-does-rewind-need-credentials.js [hunk 1]
```diff
@@ -1,11 +1,16 @@
...
+ * @param {Object} state - Global state tree
+ * @param {number|string} siteId - The unique identifier for the site
+ * @returns {boolean} Returns true if the rewind functionality is in the 'awaitingCredentials' state
+ *                    and the preflight test has failed. Otherwise, returns false.
  */
 export default function getDoesRewindNeedCredentials( state, siteId ) {
-	return [ 'awaitingCredentials' ].includes( getRewindState( state, siteId ).state );
+	const preflightStatus = getPreflightStatus( state, siteId );
+	const rewindState = getRewindState( state, siteId ).state;
+	return preflightStatus === PreflightTestStatus.FAILED && rewindState === 'awaitingCredentials';
```
Comment (Initsogar) [inline client/state/selectors/get-does-rewind-need-credentials.js]: Previously we only checked if the rewind state was `awaitingCredentials`, now we also consider if the preflight status is `failed` in order to decide if a site needs credentials or no.

File: client/components/data/query-jetpack-credentials-status/index.jsx [hunk 2]
```diff
@@ -2,10 +2,15 @@ import PropTypes from 'prop-types';
...
+import {
+	hasJetpackCredentials,
+	isRequestingJetpackCredentialsTest,
+} from 'calypso/state/jetpack/credentials/selectors';
 
 const request = ( siteId, role ) => ( dispatch, getState ) => {
-	if ( ! isRequestingJetpackCredentialsTest( getState(), siteId, role ) ) {
+	const hasCredentials = hasJetpackCredentials( getState(), siteId, role );
+
+	if ( hasCredentials && ! isRequestingJetpackCredentialsTest( getState(), siteId, role ) ) {
```
Comment (Initsogar) [inline client/components/data/query-jetpack-credentials-status/index.jsx:13]: Previously we were trying to test credentials even for sites without credentials :(

**Comment by rustamshaikh** [comment]
## Testing notes

### ✅ Self-hosted without credentials

- [x] Validate the site has no credentials configured
- [x] Ensure Jetpack Settings page says Disconnected and prompt you to enter credentials.
- [x] Ensure all restore buttons are disabled in the following pages (Calypso Blue and Jetpack Cloud):
        - `/backup/`
        - `/activity-log/`
        Screenshot 2024-01-04 at 5 57 11 PM
        Screenshot 2024-01-04 at 5 57 46 PM


- [x] Ensure you see the banner asking credentials in /backup/
   
   Screenshot 2024-01-04 at 5 59 55 PM

- [x] Ensure we are not calling `/rewind/credentials/test?role=main` endpoint when navigating to the backup page in a site without credentials.

Continuing on the next set of tests...

**Comment by rustamshaikh** [comment]
### ✅ Self-hosted with credentials - valid credentials

- [x] Validate the site has credentials configured
- [x] Ensure the credentials are valid, you can check in RWDB.
- [x] Ensure Jetpack Settings page says Connected and doesn't prompt you to enter credentials.
- [x] Ensure all restore buttons are enabled, and no banner asking credentials are presented in the following pages (Calypso Blue and Jetpack Cloud) -
         - `/backup/`
         - `/activity-log/`
         
Observation: @Initsogar  - I see it took some time and page refreshes to reflect credentials are added on Calypso blue. For me it was still showing the credentials banner on Calypso blue' backup page, I tried refreshing several times the `backup` page but it did not work, then I clicked on `Enable restores` button, and got redirected to the settings page, even there credentials were not present, then I switched back to `backup` page again and it was showing all the Restore buttons as enabled and then the settings page were also reflecting the credentials properly when revisited. Maybe some caching issue 🤔 in my browser? Not sure but just noting here :) 

Screenshot 2024-01-04 at 6 15 33 PM

**Comment by rustamshaikh** [comment]
### ✅ Self-hosted with credentials - invalid credentials

- [x] Validate the site has credentials configured
- [x] Ensure the credentials are no longer valid, you can check in RWDB.
- [x] Ensure all restore buttons are disabled in the following pages (Calypso Blue and Jetpack Cloud):
       - `/backup/`
       - `/activity-log/`
- [x] Ensure you see the banner asking credentials in `/backup/`

**Comment by rustamshaikh** [comment]
### ✅ Atomic site & Pressable

Atomic site: http://rustamwptest.wpcomstaging.com
Pressable site: http://professional-lamprey.mystagingwebsite.com

- [x] Ensure Jetpack Settings page says Connected and doesn't prompt you to enter credentials.
- [x] Ensure all restore buttons are **enabled** in the following pages (Calypso Blue and Jetpack Cloud):
        - `/backup/`
        - `/activity-log/`
        
        
 Under Atomic site & Pressable section,

> Ensure all restore buttons are **disabled** in the following pages (Calypso Blue and Jetpack Cloud)

I think you meant "enabled" :)

**Review by rustamshaikh — APPROVED** [review]
LGTM except that one observation of late updation of credentials status (maybe browser cache related)

**Comment by Initsogar** [comment]
> Observation: @Initsogar - I see it took some time and page refreshes to reflect credentials are added on Calypso blue. For me it was still showing the credentials banner on Calypso blue' backup page, I tried refreshing several times the `backup` page but it did not work, then I clicked on `Enable restores` button, and got redirected to the settings page, even there credentials were not present, then I switched back to `backup` page again and it was showing all the Restore buttons as enabled and then the settings page were also reflecting the credentials properly when revisited. Maybe some caching issue 🤔 in my browser? Not sure but just noting here :)

@rustamshaikh Interesting! I will play with this scenario a bit when I return next week. I'm guessing it is still marking that you don't have credentials because the credential tests were failing. It triggers a new test anytime you enter the backup page. Probably not directly related to this PR.

Do you remember if you saw the credentials settings changed to "Connected" when you added the credentials there?

CleanShot 2024-01-04 at 08 07 36@2x

> I think you meant "enabled" :)

Yes :D, thank you!

---

Thank you so much for your review and tests!!

**Comment by rustamshaikh** [comment]
> Do you remember if you saw the credentials settings changed to "Connected" when you added the credentials there?

I remember that it changed to Connected on Jetpack Cloud, but not on Calyspo Blue, it was showing disconnected with credentials form being empty.

**Comment by Initsogar** [comment]
> > Do you remember if you saw the credentials settings changed to "Connected" when you added the credentials there?
> 
> I remember that it changed to Connected on Jetpack Cloud, but not on Calyspo Blue, it was showing disconnected with credentials form being empty.

Ahh, okay. So you added the credentials in Jetpack Cloud, and then you wanted to validate if everything is still working as usual in Calypso Blue, but the credentials settings page shows like you don't have credentials.

Thanks again, I'm taking notes and will revisit this next week.

**Pull request merged**